micromanager_gui/main_window.py

```diff
@@ -148,92 +156,105 @@ def add_frame_explorer(self, name, array):
+            if tuple(i+1 for i in im_idx) == seq.shape:
+
+                shutil.rmtree(temp_folder.parent)
                 
-                #channels
-                if event_index_c > 0 and event_index_z == 0 and event_index_p == 0 and event_index_t == 0:
-                    empty_im = np.empty(((1,)*(len(sequence.axis_order)-2) + (z_stack_length,1,) + image.shape), dtype=np.uint16)
-                    layer.data = np.concatenate((layer.data,empty_im), axis=-3)
-
-                #xypositions
-                if event_index_p > 0 and event_index_c == 0 and event_index_z == 0 and event_index_t == 0:
-                    empty_im = np.empty(((1,)*(len(sequence.axis_order)-2) + (z_stack_length,c_stack_length,) + image.shape), dtype=np.uint16)
-                    layer.data = np.concatenate((layer.data,empty_im), axis=-5)
-
-                #timepoints
-                if event_index_t > 0 and event_index_z == 0 and event_index_c == 0 and event_index_p == 0:
-                    empty_im = np.empty(((1,p_stack_length,z_stack_length,c_stack_length,) + image.shape), dtype=np.uint16)
-                    layer.data = np.concatenate((layer.data,empty_im), axis=-6)
-
-            layer.data[event_index_t,event_index_p,event_index_z,event_index_c, ...] = image
-         
-            #set which dimension to display in napari viewer -> viewer.dims.set_point(axis, index)
-            self.viewer.dims.set_point(3, event_index_c)
-            self.viewer.dims.set_point(2, event_index_z)
-            self.viewer.dims.set_point(1, event_index_p)
-            self.viewer.dims.set_point(0, event_index_t)   
+            #save layer when acquisition is finished
+            if self.mda.save_groupBox.isChecked():
 
-        except KeyError:
+                if tuple(i+1 for i in im_idx) == seq.shape:
+
+                    name = self.viewer.layers.selection.active.name
+
+                    save_path = Path(self.mda.dir_lineEdit.text()) / name
+            
+                    self.viewer.layers[name].save(str(save_path))
+
+
+        except StopIteration:
+
+            if self.mda.save_groupBox.isChecked():
+                file_name = self.mda.fname_lineEdit.text()
+                layer_name = f"{file_name}_{datetime.now().strftime('%H:%M:%S')}"
+            else:
+                layer_name = f"mda_{datetime.now().strftime('%H:%M:%S')}"
+
+            _image = image[(np.newaxis,) * len(seq.shape)]
+            layer = self.viewer.add_image(_image, name=layer_name)
+            labels = [i for i in seq.axis_order if i in event.index] + ["y", "x"]
+
+            self.viewer.dims.axis_labels = labels
+            layer.metadata["useq_sequence"] = seq
+            layer.metadata["uid"] = seq.uid
+
+            layer = self.viewer.layers[layer_name]
+
+            # get the index of the incoming image
+            im_idx = tuple(event.index[k] for k in seq.axis_order if k in event.index)
+            
+            #save first image in the temp folder
```

Comment:
you're duplicating code from above... which is always a good sign that you should extract something into a helper function